pkg/tracking/store/sql/experiments_search_builder.go

```diff
@@ -0,0 +1,278 @@
+	filterCond      = regexp.MustCompile(`^(?:(\w+)\.)?("[^"]+"|` + "`[^`]+`" + `|[\w\.]+)\s+(<|<=|>|>=|=|!=|(?i:I?LIKE)|(?i:(?:NOT )?IN))\s+(\((?:'[^']+'(?:,\s*)?)+\)|"[^"]+"|'[^']+'|[\w\.]+)$`)
+	experimentOrder = regexp.MustCompile(`^(?:attr(?:ibutes?)?\.)?(\w+)(?i:\s+(ASC|DESC))?$`)
+)
+
+// PageToken.
+type PageToken struct {
+	Offset int32 `json:"offset"`
+}
+
+func applyExperimentsLimitFilter(query *gorm.DB, maxResults int64) (*gorm.DB, int) {
+	return query.Limit(int(maxResults) + 1), int(maxResults)
+}
+
+func applyExperimentsOffsetFilter(query *gorm.DB, pageToken string) (*gorm.DB, int, *contract.Error) {
+	var offset int
+
+	if pageToken != "" {
+		var token PageToken
+		if err := json.NewDecoder(
+			base64.NewDecoder(
+				base64.StdEncoding,
+				strings.NewReader(pageToken),
+			),
+		).Decode(&token); err != nil {
+			return nil, 0, contract.NewError(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf("invalid page_token '%s': %s", pageToken, err),
+			)
+		}
+
+		offset = int(token.Offset)
+	}
+
+	return query.Offset(offset), offset, nil
+}
+
+func applyExperimentsLifecycleStagesFilter(query *gorm.DB, runViewType protos.ViewType) *gorm.DB {
+	switch runViewType {
+	case protos.ViewType_ACTIVE_ONLY:
+		query = query.Where(
+			"lifecycle_stage IN (?)", []models.LifecycleStage{
+				models.LifecycleStageActive,
+			},
+		)
+	case protos.ViewType_DELETED_ONLY:
+		query = query.Where(
+			"lifecycle_stage IN (?)", []models.LifecycleStage{
+				models.LifecycleStageDeleted,
+			},
+		)
+	case protos.ViewType_ALL:
+		query = query.Where(
+			"lifecycle_stage IN (?)", []models.LifecycleStage{
+				models.LifecycleStageActive,
+				models.LifecycleStageDeleted,
+			},
+		)
+	}
+
+	return query
+}
+
+func applyExperimentsOrderBy(query *gorm.DB, orderBy []string) (*gorm.DB, *contract.Error) {
+	expOrder := false
+
+	for _, o := range orderBy {
+		parts := experimentOrder.FindStringSubmatch(o)
+		if len(parts) == 0 {
+			return nil, contract.NewError(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf("invalid order_by clause '%s'", o),
+			)
+		}
+
+		column := parts[1]
+		switch column {
+		case "experiment_id":
+			expOrder = true
+
+			fallthrough
+		case "name", "creation_time", "last_update_time":
+		default:
+			return nil, contract.NewError(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf(
+					"invalid attribute '%s'. Valid values are ['name', 'experiment_id', 'creation_time', 'last_update_time']",
+					column,
+				),
+			)
+		}
+
+		query = query.Order(clause.OrderByColumn{
+			Column: clause.Column{Name: column},
+			Desc:   len(parts) == 3 && strings.ToUpper(parts[2]) == "DESC",
+		})
+	}
+
+	if len(orderBy) == 0 {
+		query = query.Order("experiments.creation_time DESC")
+	}
+
+	if !expOrder {
+		query = query.Order("experiments.experiment_id ASC")
+	}
+
+	return query, nil
+}
+
+//nolint:funlen,gocognit,nestif,cyclop,goconst,mnd,forcetypeassert
```

Comment:
Not sure if I like this, can this function really not be split into helper functions?
The entire `switch entity` seems extractable? And each case in that switch can probably be a separate function as well.

Reply:
Yeah, is fine.